aioairq/core.py

```diff
@@ -128,27 +201,88 @@ async def get_latest_data(
         return data
 
     async def get(self, subject: str) -> dict:
-        """Return the given subject from the air-Q device"""
+        """Return the given subject from the air-Q device.
+
+        This function only works on a limited set of subject specified in _supported_routes.
+        Prefer using more specialized functions."""
         if subject not in self._supported_routes:
             raise NotImplementedError(
-                f"subject must be in {self._supported_routes}, got {subject}"
+                "AirQ.get() is currently limited to a set of requests, returning "
+                f"a dict with a key 'content' (namely {self._supported_routes})."
             )
 
+        return await self._get_json_and_decode("/" + subject)
+
+    async def _get_json(self, relative_url: str) -> dict:
+        """Executes a GET request to the air-Q device with the configured timeout
+        and returns JSON data as a dictionary.
+
+        relative_url is expected to start with a slash."""
+
         async with self._session.get(
-            f"{self.anchor}/{subject}", timeout=self._timeout
+                f"{self.anchor}{relative_url}", timeout=self._timeout
         ) as response:
-            html = await response.text()
+            json_string = await response.text()
 
         try:
-            encoded_message = json.loads(html)["content"]
-        except (json.JSONDecodeError, KeyError):
+            return json.loads(json_string)
+        except json.JSONDecodeError as e:
             raise InvalidAirQResponse(
-                "AirQ.get() is currently limited to a set of requests, returning "
-                f"a dict with a key 'content' (namely {self._supported_routes}). "
-                f"AirQ.get({subject}) returned {html}"
-            )
+                "_get_json() must only be used to query endpoints returning JSON data. "
+                f"{relative_url} returned {json_string}."
+            ) from e
+
+    async def _get_json_and_decode(self, relative_url: str) -> any:
+        """Executes a GET request to the air-Q device with the configured timeout
+        decodes the response and returns JSON data.
+
+        relative_url is expected to start with a slash."""
+
+        async with self._session.get(
+                f"{self.anchor}{relative_url}", timeout=self._timeout
+        ) as response:
+            json_string = await response.text()
+
+        try:
+            json_data = json.loads(json_string)
+        except json.JSONDecodeError as e:
+            raise InvalidAirQResponse(
+                "_get_json() must only be used to query endpoints returning JSON data. "
+                f"{relative_url} returned {json_string}."
+            ) from e
+
+        encoded_message = json_data["content"]
```

Comment:
I expect this may trigger a `KeyError` as, unfortunately, not all `relative_url` will return `json_data` dicts with "`content"` key (even if they are valid jsons). Which is why `get` does the explicit test of `subject` against `self._supported_routes`. I don't insist on the Look Before You Leap approach of `get` over to the EAFP approach of `_get_json_and_decode` but then maybe add a try-except with a bit more context, suggesting the supported routes. I see no better way than to just list them explicitly.

Or have you experienced a different behaviour?

Reply:
Not sure, you understand my way. I'll try to explain.

- `_get_json()` just fetches the data
- `_get_json_and_decode()` fetches and looks for the `content` key.

Both functions are private, so assuming all library authors know what they are doing, you won't run into the `KeyError`. I am aware, not every endpoint has the `content` key. See `blink()` calling `_get_json("/blink")` and `get_log()` calling `_get_json_and_decode("/log")`.

Public functions are

- Returning specify data, for example `get_possible_led_themes()`. They call `_get_json...()`.
- `get(subject)`. That's the general purpose function (you already had in place). This function checks for valid subjects and gives the user a good error message, if he calls `get("stupid-subject")`.
  - IMHO going towards going 1.0.0 I would deprecate and remove that function as well, to only provide specific functions. A general purpose GET request the user can do on his own, no need for `aioairq` in this case.

---

I did not check all the APIs air-Q offers. I assume, there could be the need to later have two different functions like

- `get_json_and_just_decode()` (when there is JSON, but no `content` key in the JSON) and
- `get_json_decode_and_return_content_key()` (for the currently supported endpoints which return `content` in the JSON)

but this is then an implementation detail and can easily be refactored, without breaking anything :-)